src/browser-polyfill.js

```diff
@@ -6,7 +6,58 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 "use strict";
 
-if (typeof browser === "undefined") {
+/**
+ * Returns true if the given object is defined and has an entry with
+ * the name `key` and the value of object[key] isn’t undefined.
+ *
+ * @param {*} object The object to test.
+ * @param {*} key The entry to test for.
+ * @returns {boolean} True if the value for obj[key] is defined.
+ */
+const isDefined = (object, key) => {
+  if (object === window && window !== this) {
+    object = this;
+  }
+  if (typeof object !== "object" || object === null) {
+    return false;
+  }
+  return key in object && typeof object[key] !== "undefined";
+};
+
+/**
+ * Returns true if the given object is an object with a `then` method, and can
+ * therefore be assumed to behave as a Promise.
+ *
+ * @param {*} value The value to test.
+ * @returns {boolean} True if the value is thenable.
+ */
+const isThenable = value => {
+  return value && typeof value === "object" && typeof value.then === "function";
+};
+
+if (typeof browser === "undefined" || !(() => {
+  // If `browser.runtime.lastError` doesn’t exist, assume promies are supported.
+  let supportsPromises = !(isDefined(window.browser, "runtime")
+    ? isDefined(window.browser.runtime, "lastError")
+    : isDefined(window.browser, "extension")
+    && isDefined(window.browser.extension, "lastError"));
+  if (supportsPromises) {
+    return true;
+  }
+  if (!supportsPromises) {
+    try {
+      supportsPromises |= isThenable(window.browser.runtime.getPlatformInfo());
+    } catch (e) {
+      try {
+        // Microsoft Edge doesn’t support `browser.runtime.getPlatformInfo()`
+        supportsPromises |= isThenable(window.browser.windows.get(browser.windows.WINDOW_ID_CURRENT));
+      } catch (e2) {
+        // Do nothing.
+      }
+    }
+  }
+  return supportsPromises;
+})()) {
```

Comment:
That's a bit wordy, how about this function instead?

```js
function supportsPromises() {
  if (typeof browser === "undefined") {
    return false;
  }
  // If `browser.runtime.lastError` doesn’t exist, assume promises are supported.
  if (browser.runtime && typeof browser.runtime.lastError !== "undefined") {
    return true;
  }
  if (browser.extension && typeof browser.extension.lastError !== "undefined") {
    return true;
  }
  try {
    return isThenable(browser.runtime.getPlatformInfo());
  } catch (error) {}
  try {
    return isThenable(browser.windows.get(browser.windows.WINDOW_ID_CURRENT));
  } catch (error) {}
  return false;
}

if (supportsPromises()) {
```

Reply:
`browser.runtime && typeof browser.runtime.lastError !== "undefined"` returns `true` if `browser.runtime.lastError` exists and `false` if it doesn’t.

Reply:
That’s a 1-char change away, the structure remains

Reply:
Just change `!` to `=`

Reply:
> consider the check failed if browser.runtime.lastError exists, but browser.extension.lastError doesn’t for whatever reason.

That’s not what the current code does. The ternary operator checks for one `runtime` and then discards `extension`

Reply:
Well, yours discards the one in `extension` if none exists in `runtime`.

Reply:
<i title="More sarcasm">Oh great</i>, and now Firefox [stopped working](https://travis-ci.org/mozilla/webextension-polyfill/builds/393441006), again.